File: surface/widgets/swiftui/components/widgets/primitives/Avatar.swift
// ============================================================
// Clef Surface SwiftUI Widget — Avatar
//
// Displays a user or entity identity as initials inside a
// bordered circular surface. When no name is provided, falls
// back to a placeholder glyph. Size affects the diameter and
// text style.
//
// Adapts the avatar.widget spec: anatomy (root, image, fallback),
// states (loading, loaded, error), and connect attributes
// (data-part, data-size, data-state) to SwiftUI rendering.
// ============================================================

import SwiftUI

// --------------- Types ---------------

enum AvatarSize: String {
    case sm, md, lg

    var diameter: CGFloat {
        switch self {
        case .sm: return 32
        case .md: return 40
        case .lg: return 56
        }
    }

    var fontSize: CGFloat {
        switch self {
        case .sm: return 12
        case .md: return 14
        case .lg: return 20
        }
    }
}

// --------------- Helpers ---------------

private func getInitials(_ name: String) -> String {
    let trimmed = name.trimmingCharacters(in: .whitespaces)
    if trimmed.isEmpty { return "?" }
    let parts = trimmed.split(separator: " ")
    if parts.count == 1 {
        return String(parts[0].prefix(1)).uppercased()
    }
    let first = String(parts.first!.prefix(1)).uppercased()
    let last = String(parts.last!.prefix(1)).uppercased()
    return "\(first)\(last)"
}

// --------------- Component ---------------

/// Avatar view that renders user/entity identity as initials
/// inside a circular surface.
///
/// - Parameters:
///   - name: Display name used to derive initials for the fallback.
///   - src: Image source URL — triggers loaded/error states.
///   - size: Size of the avatar: "sm", "md", or "lg".
///   - fallback: Custom fallback text when no name is available.
///   - showFallback: Force the fallback to display regardless of src.
struct AvatarView: View {
    var name: String = ""
    var src: String? = nil
    var size: AvatarSize = .md
    var fallback: String? = nil
    var showFallback: Bool = false

    private var displayText: String {
        fallback ?? getInitials(name)
    }

    var body: some View {
        ZStack {
            Circle()
                .fill(Color(.systemGray5))
                .overlay(
                    Circle()
                        .stroke(Color(.systemGray3), lineWidth: 1)
                )

            Text(displayText)
                .font(.system(size: size.fontSize, weight: .bold))
                .foregroundColor(.primary)
        }
        .frame(width: size.diameter, height: size.diameter)
        .accessibilityLabel(name.isEmpty ? "Avatar" : name)
    }
}

File: surface/widgets/winui/components/widgets/primitives/Avatar.cs
// ============================================================
// Clef Surface WinUI Widget — Avatar
//
// Displays a user or entity identity as initials inside a
// bordered circular surface. When no name is provided, falls
// back to a placeholder glyph. Size affects the diameter and
// text style.
//
// Adapts the avatar.widget spec: anatomy (root, image, fallback),
// states (loading, loaded, error), and connect attributes
// (data-part, data-size, data-state) to WinUI 3 rendering.
// ============================================================

using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI.Xaml.Shapes;
using Microsoft.UI;
using Windows.UI;

namespace Clef.Surface.WinUI.Widgets.Primitives;

public sealed class ClefAvatar : UserControl
{
    public static readonly DependencyProperty DisplayNameProperty =
        DependencyProperty.Register(nameof(DisplayName), typeof(string), typeof(ClefAvatar),
            new PropertyMetadata("", OnPropertyChanged));

    public static readonly DependencyProperty SizeProperty =
        DependencyProperty.Register(nameof(Size), typeof(string), typeof(ClefAvatar),
            new PropertyMetadata("md", OnPropertyChanged));

    public static readonly DependencyProperty FallbackTextProperty =
        DependencyProperty.Register(nameof(FallbackText), typeof(string), typeof(ClefAvatar),
            new PropertyMetadata(null, OnPropertyChanged));

    public string DisplayName { get => (string)GetValue(DisplayNameProperty); set => SetValue(DisplayNameProperty, value); }
    public string Size { get => (string)GetValue(SizeProperty); set => SetValue(SizeProperty, value); }
    public string FallbackText { get => (string)GetValue(FallbackTextProperty); set => SetValue(FallbackTextProperty, value); }

    private readonly Grid _root;
    private readonly Ellipse _background;
    private readonly TextBlock _initialsText;

    public ClefAvatar()
    {
        _root = new Grid();
        _background = new Ellipse
        {
            Fill = new SolidColorBrush(Colors.LightSlateGray),
            Stroke = new SolidColorBrush(Colors.Gray),
            StrokeThickness = 1
        };
        _initialsText = new TextBlock
        {
            HorizontalAlignment = HorizontalAlignment.Center,
            VerticalAlignment = VerticalAlignment.Center,
            FontWeight = Microsoft.UI.Text.FontWeights.Bold
        };
        _root.Children.Add(_background);
        _root.Children.Add(_initialsText);
        Content = _root;
        UpdateVisual();
    }

    private static void OnPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        => ((ClefAvatar)d).UpdateVisual();

    private void UpdateVisual()
    {
        var (sizeDp, fontSp) = Size switch
        {
            "sm" => (32.0, 12.0),
            "lg" => (56.0, 20.0),
            _ => (40.0, 14.0)
        };
        _root.Width = sizeDp;
        _root.Height = sizeDp;
        _background.Width = sizeDp;
        _background.Height = sizeDp;
        _initialsText.FontSize = fontSp;
        _initialsText.Text = FallbackText ?? GetInitials(DisplayName);
    }

    private static string GetInitials(string name)
    {
        if (string.IsNullOrWhiteSpace(name)) return "?";
        var parts = name.Trim().Split(' ', System.StringSplitOptions.RemoveEmptyEntries);
        return parts.Length == 1
            ? parts[0][..1].ToUpper()
            : $"{char.ToUpper(parts[0][0])}{char.ToUpper(parts[^1][0])}";
    }
}
